fadtk/model_loader.py

```diff
@@ -463,6 +463,280 @@ def float32_to_int16(self, x):
+    """
+    W2V2base model from https://huggingface.co/facebook/wav2vec2-base-960h
+
+    Please specify the layer to use (1-12).
+    """
+    def __init__(self, size='960h', layer=12, limit_minutes=6):
+        super().__init__(f"w2v2base" + ("" if layer == 12 else f"-{layer}"), 768, 16000)
+        self.huggingface_id = f"facebook/wav2vec2-base-{size}"
+        self.layer = layer
+        self.limit = limit_minutes * 60 * self.sr
+        
+    def load_model(self):
+        from transformers import AutoProcessor
+        from transformers import Wav2Vec2Model
+        
+        self.model = Wav2Vec2Model.from_pretrained(self.huggingface_id)
+        self.processor = AutoProcessor.from_pretrained(self.huggingface_id)
+        self.model.to(self.device)
+
+    def _get_embedding(self, audio: np.ndarray) -> np.ndarray:
+        # Limit to 9 minutes
+        if audio.shape[0] > self.limit:
+            log.warning(f"Audio is too long ({audio.shape[0] / self.sr / 60:.2f} minutes > {self.limit / self.sr / 60:.2f} minutes). Truncating.")
+            audio = audio[:self.limit]
+
+        inputs = self.processor(audio, sampling_rate=self.sr, return_tensors="pt").to(self.device)
+        with torch.no_grad():
+            out = self.model(**inputs, output_hidden_states=True)
+            out = torch.stack(out.hidden_states).squeeze() # [13 layers, timeframes, 768]
+            out = out[self.layer] # [timeframes, 768]
+
+        return out
+
+
+class W2V2largeModel(ModelLoader):
+    """
+    W2V2large model from https://huggingface.co/facebook/wav2vec2-large-960h
+
+    Please specify the layer to use (1-24).
+    """
+    def __init__(self, size='960h', layer=24, limit_minutes=6):
+        super().__init__(f"w2v2large" + ("" if layer == 24 else f"-{layer}"), 1024, 16000)
+        self.huggingface_id = f"facebook/wav2vec2-large-{size}"
+        self.layer = layer
+        self.limit = limit_minutes * 60 * self.sr
+        
+    def load_model(self):
+        from transformers import AutoProcessor
+        from transformers import Wav2Vec2Model
+        
+        self.model = Wav2Vec2Model.from_pretrained(self.huggingface_id)
+        self.processor = AutoProcessor.from_pretrained(self.huggingface_id)
+        self.model.to(self.device)
+
+    def _get_embedding(self, audio: np.ndarray) -> np.ndarray:
+        # Limit to 9 minutes
+        if audio.shape[0] > self.limit:
+            log.warning(f"Audio is too long ({audio.shape[0] / self.sr / 60:.2f} minutes > {self.limit / self.sr / 60:.2f} minutes). Truncating.")
+            audio = audio[:self.limit]
+
+        inputs = self.processor(audio, sampling_rate=self.sr, return_tensors="pt").to(self.device)
+        with torch.no_grad():
+            out = self.model(**inputs, output_hidden_states=True)
+            out = torch.stack(out.hidden_states).squeeze() # [25 layers, timeframes, 1024]
+            out = out[self.layer] # [timeframes, 1024]
+
+        return out
+
```

Comment:
I see that the code for W2V2 base and W2V2 large are mostly identical. Would it be better to reuse some duplicate parts of the code by using abstractions?

This also applies to the base and large variants of HuBERT, and WavLM

Reply:
Updated!
I also integrated more model types of _whisper_. Now, the most widely used speech models, i.e.,  _w2v2, hubert, wavlm_, and _whisper_  have been included, which should be sufficient for the majority of speech tasks.